first draft for forces and impulses, one enhanced demo and tests for 2D

MichaelRiss opened the pull request:
Hello Sebastien!

I send you a draft of force and impulse integration into nphysics.

Some aspects:
* Theoretically the implementation should/could already be complete for 2D and 3D. However, so far I only tested and played around in 2D. I wanted you to have a look at the design decisions first before I go into fine tuning and testing the code also in 3D.
* I looked at the construct with the ```BodyForceGenerator``` but couldn't grasp its philosophy. Therefore - for the time being - I replaced it with my own construct:
  * Forces get appended and keep acting on the bodies until they get cleared.
  * Impulses act one-time only.

  I believe this is performance-wise a good solution as forces are usually act over time and impulses are usually instantaneous. Other engines clear the forces in each step and then the application has to re-add the forces immediately afterwards. That's a lot of overhead in my opinion, especially when you think about moving the physics calculation to the GPU eventually. That would result in a lot of data traffic between CPU and GPU each simulation step.
Please comment on it and maybe explain what your design idea with the ```BodyForceGenerator``` is.
* In the functions ```append_force_wrt_point``` and ```apply_impulse_wrt_point``` I chose to let the relative pointer point from where the force acts towards the center of mass. When I imagine the vectors for the force decomposition in components that act on the center of mass and in components that make the body rotate, it is the intuitive way to do it. However, bullet - it seems - has the opposite idea: They have the pointer point from the center of mass to the point where the force acts on. What do you think? Follow intuition or follow convention and make it easy for people that switch to nphysics from other engines? I start to lean towards the bullet approach as their idea could also be seen as a "local vector" where the origin is at the center of mass of the corresponding body.
* The gravity handling has been pulled into ```RigidBody```. This design trade-off means that each ```RigidBody``` contains a copy of the gravity of its ```World```. It was either this duplication of data or a back reference from each ```RigidBody``` to its ```World``` in order to access the gravity. But this would mean that each ```World``` must be encapsulated into an ```Rc<T>``` so that we can give out ```Weak<T>``` pointers to each ```RigidBody```. I'm undecided on this, so I went with the less intrusive option. If you favor the weak pointer over the data duplication, the change shouldn't cause too much problems in my code sections.
* During the development of the tests and the enhanced demo I repeatedly ran into problems with the deactivation of resting bodies. As a user of the library I think I would prefer calls similar to PhysX where the calls for applying forces and impulses contain a boolean flag ```wakeup``` that activates the corresponding body. That might be an API change that should probably even be done before the initial merge it into the master. 
* The ```nphysics_testbed2``` modification to allow user defined key controls is just a draft. Eventually I would like to avoid the code duplication for each key. But how do you feel generally about the idea of allowing demos to provide user defined functions?
* About the tests:
  * The first test with the unrealistic big time step should probably be removed.
  * Right now each test routine tests quite a range of functionality. What do you think about cutting these large test routines down so that each tests just a single feature? I would lean to it, but it would lead to code duplication ... and probably a more impressive looking test output. :smiley: 
  * While writing the tests I stumbled across two bugs that seem to be situated in other parts of nphysics or rust.
    * One is a division by zero panic when instantiating two bodies at exactly the same position and then starting the simulation and applying forces on them. 
    * The other happens when you remove a ```RigidBody``` from ```World```.

    I left the commented sections that trigger the bugs in the tests.rs file for further investigation.
* I inverted the scroll wheel direction in ```nphysics_testbed2``` so that it's more in line with the zoom scroll direction of applications like google earth, etc.
* The rotation direction in ```nphysics_testbed2``` may be counter intuitive as it seems the y-axis of sfml points downwards and therefore the z-axis (rotation axis) points into the monitor. Therefore the bodies on the screen rotate in the opposite direction than what the right hand three finger rule would intuitively suggest.

I hope this draft is of some use to you and does not go completely against your design ideas. And if so, just tell me, I know that I took a risk with this design draft. Also, don't feel pressured to review it quickly, it took a long time to code, it can take a long time to review then :laughing: and I'm not in a rush.
Have some fun with the "meteor impacts" in the demo when switching off the forces after the bodies have reached some altitude!

cu
Michi

sebcrozet commented:
Hi!

So far I just played with the demo and this looks pretty cool (especially when I press `3`). I'll take a deeper look at your comment and your code within the next few days!

sebcrozet commented:
Sorry if I waited too long! Got many things to deal with lately so I kind of delayed this review over and over.

So I will first answer your questions/comment on your design choices. I won't comment the tests for now as this is not the most important.


> I looked at the construct with the BodyForceGenerator but couldn't grasp its philosophy.

The `BodyForceGenerator` construct actually comes from an older design of `nphysics`. At that time, the idea of allowing user-defined forces was designed in that way: the `World` contained a list of objects that implement the `Integrator<RigidBody>` trait that are responsible for the update of individual rigid bodies during the first phase of the physics pipeline. Hence, if the user needed to add its own force generator, he only had to create its own implementor of `Integrator<...>` and add it to the world. I removed this because it caused performances issues and was not so simple to use. Though I kept the `BodyForceGenerator` to handle the gravity.


> I believe this is performance-wise a good solution as forces are usually act over time and impulses are usually instantaneous. Other engines clear the forces in each step and then the application has to re-add the forces immediately afterwards. That's a lot of overhead in my opinion, especially when you think about moving the physics calculation to the GPU eventually. That would result in a lot of data traffic between CPU and GPU each simulation step.

I am actually not sure why physics engines usually clear forces at each time steps. One guess would be that as their variables accumulate forces from multiple sources, they must clear them at the beginning of the timestep to be sure to take in account modified force values like e.g. the gravity. In any cases, I don't think that the GPU-related argument is relevant here as I think a really efficient GPU implementation would require a memory layout much different from the current one (for example it would be more beneficial to have the force accumulator of every rigid body grouped on a single large array the rigid body itself would point into so that the GPU can perform integration operation on those forces in a more memory-efficient manner); though I never studied the question of GPU-based physics engine so I can't do anything more that speculations here.

In any case, I am in favor of keeping a force accumulator that is not cleared at each update. 

The design I had in mind for this is that user-defined forces would come from some objects that would look pretty much like the current joints. Currently, to couple the behavior of two rigid bodies, you have to add a joint to the scene. In a similar fashion, to add permanent forces to a rigid body you could add to the scene an object that is responsible for this force generation; such object would be a "motor" that acts unilaterally on the RigidBody instead of on a couple of rigid bodies like joints.


> In the functions append_force_wrt_point and apply_impulse_wrt_point I chose to let the relative pointer point from where the force acts towards the center of mass. When I imagine the vectors for the force decomposition in components that act on the center of mass and in components that make the body rotate, it is the intuitive way to do it. However, bullet - it seems - has the opposite idea: They have the pointer point from the center of mass to the point where the force acts on. What do you think? Follow intuition or follow convention and make it easy for people that switch to `nphysics` from other engines? I start to lean towards the bullet approach as their idea could also be seen as a "local vector" where the origin is at the center of mass of the corresponding body.

Let a force be defined by its point of action, its direction, and its magnitude.

You approach, while somehow intuitive from the force-action point of view, does not seem to be a viable answer to the more mathematically rigorous question « What are the basis in which are expressed each of my vectors? » (I am talking of vector space basis here). To answer this question, there are actually several incompatible « conventions ». The two most common are :
  1. Express both the point of action and the direction in the rigid body's local frame. This is probably what Bullet does (you get the actual application point and its direction in world-space coordinates by multiplying them by the rigid body transformation matrix).
  2. Express the point of action in the rigid body's local frame, but keep the direction in world-space coordinates. While this might seems weird, this is very useful when you want to keep a force acting on the same direction without worrying about the object orientation over time.


> The gravity handling has been pulled into RigidBody. This design trade-off means that each RigidBody contains a copy of the gravity of its World. It was either this duplication of data or a back reference from each RigidBody to its World in order to access the gravity. But this would mean that each World must be encapsulated into an Rc<T> so that we can give out Weak<T> pointers to each RigidBody. I'm undecided on this, so I went with the less intrusive option. If you favor the weak pointer over the data duplication, the change shouldn't cause too much problems in my code sections.

Given your other design choices, you did the right thing here. Putting the `World` on an `Rc<...>` is not acceptable as this is prone to bad design in the future (i-e tendency to share the world all over the place for futile reasons just like many other physics engines do).
Though I don't think pulling the gravity as a field of the `RigidBody` (together with the "has_gravity") is a very aesthetic solution. If a force applies to some rigid bodies but not others, it should not be called "gravity". It is some weird global force that follows some user-defined rules, but this is not the gravity as we know/expect it. So I'd prefer to let the gravity remain on the world. If the user wants to cancel it for some reasons he is free to apply an opposite force to the specific objects.


> During the development of the tests and the enhanced demo I repeatedly ran into problems with the deactivation of resting bodies. As a user of the library I think I would prefer calls similar to PhysX where the calls for applying forces and impulses contain a boolean flag wakeup that activates the corresponding body. That might be an API change that should probably even be done before the initial merge it into the master.

Yeah, rigid body deactivation is always tricky to handle. The thing is, an object is not able activate itself properly (that's why the rigid body's `.activate(...)` method is hidden from the documentation). The only proper way to activate a rigid body is to call the world's `ActivationManager`'s `.will_activate(...)` method. This cannot be done by the user currently. So there are several exclusive options here:
  * Add a method to the `World` to let the user get a mutable reference to the `ActivationManager` so that he can call the relevant method whenever he wishes.
  * Add a method `.request_activation(...)` to the world that will call the relevant method from the `ActivationManager`.
  * Add a boolean to the rigid bodies. If this boolean is `true` at the beginning of the world update, then the world calls the `.request_activation(...)` method and resets the boolean back to `false`.

I am not sure which approach is the best. Though I am not fond of the first one because a mutable reference like this gives too much power to the user. The third option allows the design you suggested regarding the Boolean to wake up objects on force application.

So, to conclude, I am willing to merge your pull request (at least the non-test part at the moment) as a better option (see the "motor" thing a few paragraphs above) is much more complicated and won't be implemented for a while. Before a merge, I believe the following modifications are desirable (if you think they are not do not hesitate to argue):
  * Express the point of application of your force in the Rigid-Body local frame (i-e just like bullet does) so that we line up with what is in my opinion a more mathematically intuitive behavior.
  * Move the gravity back to the world (to the `BodyForceGenerator` actually) and remove the `has_gravity` flag. A force that does not affect every rigid body does not deserve the `gravity` nickname.

I just hope I did not wait to long to give you this feedback!

MichaelRiss commented:
Hi Sebastién!

Thanks for your answer and I fully understand when answering takes a bit longer, I'm busy , too.

- I will change the force/impulse wrt point to use local frame coordinates - it's clearly the better approach. In the future we could probably think about additional utility functions similar to the ```*Ext```-PhysX ones: https://developer.nvidia.com/sites/default/files/akamai/physx/Manual/RigidDynamics.html
But let's concentrate on the core functions first, I'd say.
- I will remove ```has_gravity```, you are right, optional gravity is over-engineering.
- But I'm not entirely happy about the ```BodyForceGenerator``` implementation. It requires to copy the gravity value for each time step into all active bodies again and again. It seems that there is a structurally similar problem with the activation of sleeping bodies. In both cases there are data structures stored in world that would be needed for certain function calls on rigid bodies. Currently I see  a couple of solutions to this:
  - Your third idea to the activation problem. Calling the functions on the rigid body and setting a flag in the rigid body. Later - during world update - iterate through all rigid bodies and execute the actions that need access to world data structures.
  - Add a mandatory ```world``` argument to rigid body functions that need to modify world.
  - Modify ```world.add_body()```: Instead of returning a ```Rc<RefCell<RigidBody>>``` it would return a glue object that contains handles to the rigid body and a (weak) reference to a sub-component of world. Now you can call functions on the handle that require access to both rigid body and world data structures.
  - The idea I like most (in the long term for GPU implementation) is a modification of the first idea. But instead of setting a flag in the rigid body, the function writes out an entry into a TODO-list. The TODO-list is a sub-component of world, referenced via a weak reference. During world update the TODO-list is read and the actions are executed. This way you don't have to iterate through all rigid bodies for each time step and only execute the actions that are necessary. Also, the TODO-list can be used for a future GPU-based implementation as a work package that is moved over to the GPU (and yes, the current data structures are not GPU compatible, but I want to design the API in a way that it's also efficient with a potential future GPU backend).

For the moment I think it's best when I revert the implementation to the original ```BodyForceGenerator``` implementation and then do potentially a further change if one of the above options seems suitable.
But I will need some time, me too, I'm busy with other things. ;)

Regards,
Michi

PS. I'm still thinking about your "motor" object idea for forces. Maybe something can be done about it.

MichaelRiss commented:
Hello Sébastien,

sorry, it took a bit longer here, too.

I have now
- removed the ```has_gravity``` flag
- changed the ```*_wrt_point``` functions so that the position vector points from center of mass to point of action
- reactivated the ```BodyForceGenerator``` and restriced the application to the linear part of the force.
- commented out ```world::angular_acceleration``` and ```world::set_angular_acceleration``` as I guess the functions may be superfluous. Please check and tell me if my guess is right.

I was brooding over the ```BodyForceGenerator```, there may be an alternative/probably more elegant solution: We could hand down ```gravity``` from ```world``` over ```BodySmpEulerIntegrator``` to ```euler::semi_implicit_integrate``` and add it there to ```l_acc```. This way ```gravity``` does not need to be copied into a field of ```RigidBody```.


cu
Michi

PS. Sorry for putting the accent of your name in the wrong position in the last post (corrected).

sebcrozet commented:
Hi!

> In the future we could probably think about additional utility functions similar to the *Ext-PhysX ones: https://developer.nvidia.com/sites/default/files/akamai/physx/Manual/RigidDynamics.html

Well I don't know yet my position wrt. to the addition of this kind of helper functions. I am usually not fond of adding lots of non-essential methods where the maths behind them is not that difficult for the user to write (in the end, this is all just a matter of changing the coordinate system).

> * Your third idea to the activation problem. [...] flag [...]

When I read my idea again, I don't think it is very good. Adding a flag feels more like a workaround than a solution. It is not as generalizable as the third idea.

> * Add a mandatory world argument to rigid body functions that need to modify world.
> * Modify world.add_body(): Instead of returning a Rc<RefCell<RigidBody>> it would return a glue object that contains handles to the rigid body and a (weak) reference to a sub-component of world. Now you can call functions on the handle that require access to both rigid body and world data structures.

Right now I don't have a very strong position for or against those two options. The first one seems to give too much power to a rigid body (it could theoretically do whatever it wants to the whole world. I feel uncomfortable with this even if we − the developers but not the users − are the ones that control what the bodies actually do with the world).
And at the API level, the second option would have to be presented to the user in an elegant way (and I don't think this is simple).

> * The idea I like most (in the long term for GPU implementation) is a modification of the first/third idea. But instead of setting a flag in the rigid body, the function writes out an entry into a TODO-list. The TODO-list is a sub-component of world, referenced via a weak reference. During world update the TODO-list is read and the actions are executed. This way you don't have to iterate through all rigid bodies for each time step and only execute the actions that are necessary. Also, the TODO-list can be used for a future GPU-based implementation as a work package that is moved over to the GPU (and yes, the current data structures are not GPU compatible, but I want to design the API-calls in a way that it's also efficient with a potential future GPU backend).

Correct me if I'm wrong, but this looks a lot like some kind of event-based system. What you call a TODO-list is really an event pool that the world has to interpret/empty at each update. Among the three, I prefer this approach too though I have the feeling that a robust implementation could have some rough edges (like, what if the user activates then deactivate the rigid body in-between two world updates? We could either interpret them as being a no-op [the event and its anti-event annihilate each other), or we could interpret them individually, which may trigger other activation events for objects in contact with the rigid body in question). The other reason why I like this approach is that the current object addition/removal kind of work this way already (the objects are not added/removed right away but added to a list so that the world can deal with this at the next update).

In any case, this is a complicated topic and will require its own issue.

> I was brooding over the BodyForceGenerator, there may be an alternative/probably more elegant solution: We could hand down gravity from world over BodySmpEulerIntegrator to euler::semi_implicit_integrate and add it there to l_acc. This way gravity does not need to be copied into a field of RigidBody.

No, I'd prefer not to touch the integrator for this. The integrator should remain what it is: an ODE integrator for motion dynamics. It doesn't have to know anything as high level as what the gravity is to perform its task.


Anyway, I believe this is ready for a partial merge. Within the next few days I will still have to read the code once again and perhaps apply some modifications to the examples at merge-time but this seems globally OK for a first approach.

MichaelRiss commented:
> Correct me if I'm wrong, but this looks a lot like some kind of event-based system. What you call a TODO-list is really an event pool that the world has to interpret/empty at each update. 

Correct, that's the idea.

> ... though I have the feeling that a robust implementation could have some rough edges (like, what if the user activates then deactivate the rigid body in-between two world updates? We could either interpret them as being a no-op [the event and its anti-event annihilate each other), or we could interpret them individually, which may trigger other activation events for objects in contact with the rigid body in question).

I'm in favor of a no-op interpretation. All the state changes that the user requests happen in an instant (zero time passes). Time only passes when ```world::step()``` is called and then bodies move, bump into other bodies, activate them, etc.

I agree with all your other comments.

sebcrozet commented:
It is merged!

In the process I made a few coding-style-related modifications, moved you 2D demo to another file named `manual_forces.rs` (and reverted `convex_vee.rs` to what it was before), and modified your callback system so that the user may assign up to 9 callbacks to the 2D testbed (I unified the `add_callback{1234}` methods into a single one).

Thanks for all your efforts.
